cfg_mgmt/github.py

```diff
@@ -0,0 +1,326 @@
+import dataclasses
+import datetime
+import logging
+import os
+import typing
+import yaml
+
+import git
+from Crypto.PublicKey import RSA
+
+import ccc.github
+import cfg_mgmt.model as cmm
+import cfg_mgmt.util
+import ci.util
+import concourse.util
+import gitutil
+import model
+import model.github
+
+from cfg_mgmt.model import (
+    CfgQueueEntry,
+    CfgTarget,
+    CfgStatus,
+    cfg_queue_fname,
+    cfg_status_fname,
+)
+from model.github import (
+    GithubConfig,
+)
+
+
+ci.log.configure_default_logging()
+logger = logging.getLogger(__name__)
+
+
+def rotate_github_config(
+    cfg_dir: str,
+    cfg_element: model.github.GithubConfig,
+    git_helper: gitutil.GitHelper,
+    target_ref,
+    cfg_metadata,
+):
+    logger.info(f"Rotating github config '{cfg_element.name()}'")
+
+    cfg_queue = _read_config_queue(cfg_dir=cfg_dir, queue_file_name=cfg_queue_fname)
+
+    cfg_factory = model.ConfigFactory.from_cfg_dir(cfg_dir=cfg_dir, disable_cfg_element_lookup=True)
+    local_sources = cfg_mgmt.util.local_cfg_type_sources(
+        cfg_element=cfg_element,
+        cfg_factory=cfg_factory,
+    )
+
+    if len(local_sources) != 1:
+        raise NotImplementedError(
+            'Do not know how to rotate github configs sourced from more than one file.'
+        )
+
+    source_file = next((s for s in local_sources))
+
+    known_github_configs: typing.Iterable[GithubConfig] = [
+        c for c in cfg_factory._cfg_elements(
+            cfg_type_name=cfg_element._type_name,
+        )
+    ]
+
+    cfg_name = cfg_element.name()
+    if not (cfg_to_rotate := next(
+        filter(lambda c: c.name() == cfg_name, known_github_configs),
+        None,
+    )):
+        raise RuntimeError(
+            f"Did not find requested config '{cfg_name}' in config dir at '{cfg_dir}'"
+        )
+
+    old_keys = [{
+        'name': u.username(),
+        'public_key': update_user(cfg_to_rotate, u.username()),
+    } for u in cfg_to_rotate._technical_users()]
+
+    _write_github_configs(
+        cfg_dir=cfg_dir,
+        cfg_file_name=source_file,
+        github_configs=known_github_configs,
+    )
+
+    _update_config_queue(
```

Comment:
naming-nit: `_add_to` (or append, if you prefer - choose a verb matching the semantics more specifically than update, please)